translators/src/Print/Agda.hs

```diff
@@ -1,128 +1,165 @@
+printType (Con t) = pretty t
+printType (Arr t1 t2) = printType t1 <+> arr <+> printType t2
+printType (TVar t) = pretty t
+printType (PCon name types) = pretty name <+> hsep (map printType types)
+printType (DCon name [] exprs) = -- For dependent type constructors
+    pretty name <+> hsep (map printExpr exprs)
+printType (DCon name types exprs) = -- For dependent type constructors
+    pretty name <+> hsep (map printType types) <+> hsep (map printExpr exprs)
+printType (Index names ty) = braces $ typeAnn (hsep $ map pretty names) (printType ty)
 printType (Embed e) = printExpr e
 
 -- Print expressions
-printExpr :: Expr -> String
-printExpr (Constructor name) = name
-printExpr (Var var) = var
-printExpr (Nat n) = show n
-printExpr (String str) = quote str
+printExpr :: Expr -> Doc ann
+printExpr (Constructor name) = pretty name
+printExpr (Var var) = pretty var
+printExpr (Nat n) = pretty n
+printExpr (String str) = dquotes $ pretty str
 printExpr (Paren e) = parens $ printExpr e
-printExpr (Bin op e1 e2) = printExpr e1 ++ printOp op ++ printExpr e2
+printExpr (Bin op e1 e2) = printExpr e1 <+> printOp op <+> printExpr e2
 printExpr (Let ds expr) = 
-  "let\n    " ++ (intercalate "\n    " (map printDef ds)) ++ " in\n    " ++ printExpr expr
-printExpr (If cond thn els) = "if " ++ printExpr cond ++ " then " ++ printExpr thn ++ " else " ++ printExpr els
-printExpr (Where expr ds) = printExpr expr ++ "\n    where " ++ intercalate "\n    "  (map printDef ds)
-printExpr (FunCall fun args) = fun ++ " " ++ unwords (map printExpr args) -- Added case for FunCall
-printExpr (VecE []) = "[]"
-printExpr (VecE l@(_:_)) = parens $ intercalate " ∷ " (map printExpr l) ++ " ∷ []"
-printExpr (ListE []) = "[]"
-printExpr (ListE l@(_:_)) = parens $ intercalate " ∷ " (map printExpr l) ++ " ∷ []"
-printExpr (Suc t) = parens $ "suc " ++ printExpr t
-
-printOp :: Op -> String
-printOp Plus = " + "
+  "let" <+> align (vcat (map printLocalDefn ds) <+> "in") <> line <>
+  printExpr expr
+printExpr (If cond thn els) =
+  "if" <+> printExpr cond <+> "then" <+> printExpr thn <+> "else" <+> printExpr els
+printExpr (Where expr ds) =
+  printExpr expr <> line <>
+  indent 4 ("where" <> vcat (map printLocalDefn ds))
+printExpr (FunCall fun args) = pretty fun <+> (fillSep (map (group . printExpr) args))
+printExpr (VecE l) = parens $ encloseSep emptyDoc (space <> lcons <+> lbracket <> rbracket)
+  (space <> lcons <> space) (map printExpr l)
+printExpr (ListE l) = parens $ encloseSep emptyDoc (space <> lcons <+> lbracket <> rbracket)
+  (space <> lcons <> space) (map printExpr l)
+printExpr (Suc t) = parens $ "suc" <+> printExpr t
+
+printOp :: Op -> Doc ann
+printOp Plus = "+"
+
+printLocalDefn :: LocalDefn -> Doc ann
+printLocalDefn (LocDefFun var ty args expr) =
+   typeSig <> tvar <+> assign <+> align (printExpr expr)
+    where
+        typeSig = case ty of
+            Just t -> typeAnn (pretty var) (printType t) <> line
```

Comment:
Hard line break here.

We should probably add some tests that make sure we handle line breaking well: it's too easy to miss otherwise.